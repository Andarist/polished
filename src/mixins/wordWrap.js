// @flow
import messageHandlers from '../internalHelpers/_messageHandlers'

type WrapKeywords = 'break-all' | 'break-word' | 'normal'

const wrapKeywords = ['break-word', 'normal']

/**
 * Provides an easy way to change the `wordWrap` property.
 *
 * @example
 * // Styles as object usage
 * const styles = {
 *   ...wordWrap('break-word')
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   ${wordWrap('break-word')}
 * `
 *
 * // CSS as JS Output
 *
 * const styles = {
 *   overflowWrap: 'break-word',
 *   wordWrap: 'break-word',
 *   wordBreak: 'break-all',
 * }
 */

function wordWrap(wrap?: WrapKeywords | string = 'break-word') {
  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'production') {
    if (
      messageHandlers('mixins/wordWrap.js', {
        // eslint-disable-next-line prefer-rest-params
        arrityCheck: { args: arguments, max: 1 },
        typeChecks: { param: wrap, type: 'enumerable', map: wrapKeywords },
      })
    ) {
      return {}
    }
  }

  const wordBreak = wrap === 'break-word' ? 'break-all' : wrap
  return {
    overflowWrap: wrap,
    wordWrap: wrap,
    wordBreak,
  }
}

export default wordWrap
