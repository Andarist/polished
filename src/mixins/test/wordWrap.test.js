// @flow
import wordWrap from '../wordWrap'

describe('wordWrap', () => {
  beforeAll(() => {
    global.console = {
      error: jest.fn(),
      warn: jest.fn(),
    }
  })

  afterEach(() => {
    // eslint-disable-next-line no-console
    console.error.mockClear()
    // eslint-disable-next-line no-console
    console.warn.mockClear()
  })

  it('should accept break-word and properly set word-break to break-all', () => {
    expect({
      ...wordWrap('break-word'),
    }).toMatchSnapshot()
  })

  it('should accept normal as a value', () => {
    expect({
      ...wordWrap('normal'),
    }).toMatchSnapshot()
  })

  it('should default wrap to break-word', () => {
    expect({
      ...wordWrap(),
    }).toMatchSnapshot()
  })

  it('should throw an error when called with a non-string value', () => {
    wordWrap(1)
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when called with an invalid string value', () => {
    wordWrap('break-all')
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw a warning when called with more than 1 parameter', () => {
    wordWrap('break-all', 2)
    // eslint-disable-next-line no-console
    expect(console.warn.mock.calls).toMatchSnapshot()
  })

  it('should throw an error and a warning when called with multiple non-string values', () => {
    wordWrap(1, 2)
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
    // eslint-disable-next-line no-console
    expect(console.warn.mock.calls).toMatchSnapshot()
  })
})
