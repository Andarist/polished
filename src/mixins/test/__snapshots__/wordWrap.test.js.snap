exports[`wordWrap should accept break-word and properly set word-break to break-all 1`] = `
Object {
  "overflowWrap": "break-word",
  "wordBreak": "break-all",
  "wordWrap": "break-word",
}
`;

exports[`wordWrap should accept normal as a value 1`] = `
Object {
  "overflowWrap": "normal",
  "wordBreak": "normal",
  "wordWrap": "normal",
}
`;

exports[`wordWrap should default wrap to break-word 1`] = `
Object {
  "overflowWrap": "break-word",
  "wordBreak": "break-all",
  "wordWrap": "break-word",
}
`;

exports[`wordWrap should throw a warning when called with more than 1 parameter 1`] = `
Array [
  Array [
    "%c ✨ WARNING ✨ ---- mixins/wordWrap.js --%c

wordWrap %cexpects a maximum of %c1 parameter%c. However, you passed %c2 parameters%c. This additional parameter was ignored.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#wordWrap%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: green",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: goldenrod",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`wordWrap should throw an error and a warning when called with multiple non-string values 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/wordWrap.js --%c

wordWrap %creceived an enumerable value(1) that was not one of the available options.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#wordWrap%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`wordWrap should throw an error and a warning when called with multiple non-string values 2`] = `
Array [
  Array [
    "%c ✨ WARNING ✨ ---- mixins/wordWrap.js --%c

wordWrap %cexpects a maximum of %c1 parameter%c. However, you passed %c2 parameters%c. This additional parameter was ignored.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#wordWrap%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: green",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: goldenrod",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`wordWrap should throw an error when called with a non-string value 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/wordWrap.js --%c

wordWrap %creceived an enumerable value(1) that was not one of the available options.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#wordWrap%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`wordWrap should throw an error when called with an invalid string value 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/wordWrap.js --%c

wordWrap %creceived an enumerable value(break-all) that was not one of the available options.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#wordWrap%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;
